### PR TITLE
support add-argument action

### PR DESCRIPTION
**Pull request by santiweight**
This action supports an action "Add argument `xxx` to function" whenever there is an undefined variable error:

```Haskell
foo = unknown
-- becomes
foo unknown = unknown
```

The only minor wrinkle is that

```Haskell
foo = True
   where
      bar = unknown
-- becomes
foo unknown = True
   where
      bar = unknown
```

But this is good enough for now as a matter of pragmatism.

As part of this MR, I fixed some behavior. Previously an unknown-type undefined variable did not offer a "define x" option, but now does:

```Haskell
foo = unknown
-- previously no code actions
-- now becomes
foo = unknown

unknown :: _
unknown = _
```

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/3149"></a>

### CODE REVIEW

**Comment by kokobd** [comment]
Thanks for your contribution!

For reference, below are the behavior of some other language servers/IDEs on this:
- rust-analyzer in VSCode: no quickfix
- TypeScript in VSCode: no quickfix
- Java in Intellij IDEA Community: has this quickfix, but provides a dialog to modify function signature
<details>
<summary>screenshots</summary>
image
image
image
image
</details>

I'm not quite sure about the design, maybe someone else will provide better insight.

**Comment by santiweight** [comment]
Thanks for the screenshots! Very helpful context :) You note that Rust/TS have no quick fix. I assume you mean this as a positive right?

I am quite familiar with the IntelliJ interface from my previous job. I would love to support this workflow, but after I looked around the HLS and LSP issues I got the impression that such a pretty workflow would require too much work to justify. Is that you guys' experience? I really think our UX could be best in class given Haskell's type system, but I'm concerned about maintenance hell.

On a side note, it would be really great to have some sort of function:

```Haskell
alterAllUsages :: {- function -} RdrName -> (LHsExpr -> TransformT m LHsExpr) -> IDEState -> TransformT m IDEState
```

The idea being that when I add a function parameter to a definition, I can insert appropriate holes in all usages project wide. Does anyone have any thoughts on the feasibility of this? @OliverMadine seems like a candidate for such knowledge!

```Haskell
module One where

import Two

bar = foo True
----
module Two where

foo True = unknown False

----- BECOMES -----
module One where

import Two

bar = foo True _unknown
----
module Two where

foo True unknown = unknown False
```

**Comment by OliverMadine** [comment]
> The idea being that when I add a function parameter to a definition, I can insert appropriate holes in all usages project wide. Does anyone have any thoughts on the feasibility of this? @OliverMadine seems like a candidate for such knowledge!

Yes, it seems very feasible to me!
Most of the logic could be extracted from the rename plugin. It would, however, be subject to the same multi-component limitation of the rename plugin (#2193)

File: plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs
Comment (kokobd) [inline plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs:657]: Maybe you can use [`realSrcSpanToRange`](https://github.com/haskell/haskell-language-server/blob/163ebfe9d02d5aa9f2c83ef8d9db2e91cde31032/ghcide/src/Development/IDE/GHC/Error.hs#L76) and [`isSubrangeOf`](https://github.com/haskell/lsp/blob/a41954e356117d8f5cb64b20e1a9d32e8fc6a884/lsp-types/src/Language/LSP/Types/Location.hs#L89) ?

File: plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs
Comment (pepeiborra) [inline plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs:444]: could you add Haddock docs for this and other new functions please?

File: plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs
Comment (pepeiborra) [inline plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs:663]: Searching for this type signature (flipped) within the codebase, I find 3 functions that seem to duplicate this logic.

Worth reusing one of them? 

Screenshot 2022-09-18 at 16 26 18

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (pepeiborra) [inline plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:101]: You need to reformat this with stylish-haskell to pass the pre-commit-hook check

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (pepeiborra) [inline plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:931]: Avoid throwing an `error` here, as it will at best kill all other code actions and at worst kill all other plugins. Ideally you should return a `ResponseError` but that probably requires too much refactoring.

File: plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs
Comment (pepeiborra) [inline plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs:448]: Why does `f` return a `Maybe`? Couldn't it simply return its input untouched?

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (pepeiborra) [inline plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:930]: It would be sweet if the type signature was also extended if present

**Review by pepeiborra — CHANGES_REQUESTED** [review]

**Comment by drsooch** [comment]
There is a plug-in that attempts to update signatures when it can. `change-type-signature` It's pretty rudimentary and only works in a handful of cases where GHC provides enough information. When the upstream `lsp` package updates to the lsp spec 3.17. We should be able to get direct access to GHCs Error Message API and potentially be worthwhile to this case.

File: plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs:663]: I pulled this out into Development.IDE.GHC.Error, where similar functions are.

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:930]: I totally agree. I've left a TODO above for the time being, but it'll be my priority after I finish this and a follow-up action.

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:101]: Let's see if I was able to fix all the formatting issues...

File: plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs:448]: This was all a little silly. There was no need for the Maybe at all, so I reworked the passed-in function.

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:931]: I'm unclear on what to do here. For the time being I used a trace call and returned no diffs. Is ResponseError part of the MonadLSP interface?

File: plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs]: Added some comments on both

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:931]: I was able to implement adding to the type signature here: https://github.com/santiweight/haskell-language-server/tree/add-arg-type-sig

It is rather complicated... consider altering multi-type-sig-decls of the form `foo, bar :: Int`. You have to split up the declarations.

I was able to make it work, but it'll take a fair amount of testing, and so I'd much rather keep this PR as an intermediate step

**Comment by santiweight** [comment]
@drsooch I've thought about the change-type-sig plugin in this context. I think they're fairly different situations. In change-type-sig, I can only see occurrences where the new type signature is given. 

However, in this case, we do not have the full type signature: we only have the type of the argument we're adding. That means we do have to manually alter the HsSigType with the exact print api, which is far more complicated.

I implemented a working first attempt (it's surely very buggy for non-trivial layouts): https://github.com/santiweight/haskell-language-server/tree/add-arg-type-sig

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (fendor) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs]: Yes it is :)

Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs]: I summoned up the courage to reveal ResponseErrors in the CodeAction API. The only hold up is I don't know how to throw/report ResponseErrors with MonadLSP after some hoogling.

Anyone have any pointers? 😅 Note that I am returning a list of errors _and_ successes (since there are many potential code actions), so it should just be `logResponseError :: MonadLSP m => ResponseError -> m ()`, not throw

Comment (pepeiborra) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs]: I think it's fine to just use `logError` or `logWarn` and return nothing

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (pepeiborra) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:931]: Fair

**Review by pepeiborra — APPROVED** [review]
Looks good to me modulo the build failures

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs]: After some investigation, it appears that the only Recorder available to CodeActions is `Recorder (WithPriority E.Log)`, which is the logging from `Development.IDE.Core.Shake`. 

Which logger should I be using for code actions? I can't find a Logger or nonShake recorder in the parent functions...

File: plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs]: I forgot to respond. This worked perfectly thanks :)

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (pepeiborra) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs]: This is because this plugin was only recently extracted from ghcide core. The plugin needs to define its own Logger I guess.

**Comment by michaelpj** [comment]
Can you add yourself to `CODEOWNERS`? You might also want to mention this in `features.md`.

**Comment by santiweight** [comment]
It seems that the tests didn't run as part of CI? Besides that I've added myself to codeowners but I don't have write access, which is causing an error in the code owners file!

I'm happy to merge whenever now :) Thanks for the help getting this one through!

**Comment by michaelpj** [comment]
> It seems that the tests didn't run as part of CI?

They should have, we're testing `hls-refactor-plugin` which should include the new stuff, right?

> Besides that I've added myself to codeowners but I don't have write access, which is causing an error in the code owners file!

I've invited you to the repo :)

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (michaelpj) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:894]: Also, this module is also quite large, perhaps the add-action stuff could go in a separate module also?

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (michaelpj) [inline plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:960]: are you going to do the TODO in this PR?

**Review by michaelpj — APPROVED** [review]
I took a fairly superficial look, but I think we can get this over the line :)

Thank you for digging into the code actions in `hls-refactor-plugin`. As Pepe mentioned, they only recently got extracted from `ghcide`, so they quite probably be tidied up more to be more in keeping with the style of more recent code action handlers if you felt moved to!

**Comment by santiweight** [comment]
Thanks for the review @michaelpj 

A heads up on the status of this PR... I'm pretty happy with where it is right now, but I'm a little worried about my approach to using the Exactprint API, and so I'm going to attempt reimplementing the code in a better style (by following the examples in ghc-exactprint).

This MR isn't forgotten, just hopefully being replaced by something better...

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:960]: no - it's upcoming in a future PR :)

File: ghcide/src/Development/IDE/GHC/Error.hs
Comment (santiweight) [reply ghcide/src/Development/IDE/GHC/Error.hs:125]: I returned a Nothing and added a comment; I think it's clearer now...

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:894]: I actually already did this in a followup MR. Would you be okay with following up with this change (to avoid unnecessary conflicts)?

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:926]: Revised

File: plugins/hls-refactor-plugin/test/Main.hs
Comment (santiweight) [reply plugins/hls-refactor-plugin/test/Main.hs:2169]: I have already done this in a  followup change. Unfortunately fixing this now would cost me some time with conflicts, but I promise to follow up :)

**Comment by santiweight** [comment]
I was hoping to upstream some of the changes here in ghc-exactprint before merging, but I think that process will take some time.

I've answered the comments from reviewers, and I now think this is ready to merge!

**Comment by michaelpj** [comment]
I'm happy for you to continue improving things, especially since it sounds like you're definitely continuing to work in this area :)

**Pull request merged**
